_posts/2018-02-06-nova-neutron.md

```diff
@@ -0,0 +1,43 @@
+---
+author: <a href="https://twitter.com/IgnasiBarrera">Ignasi Barrera</a>
+comments: true
+date: 2018-02-06 07:00:00+00:00
+layout: post
+slug: nova-neutron
+title: OpenStack Nova and Neutron
+---
+
+One of the limitations of the jclouds implementation of the OpenStack Nova API is that is was not able to directly talk to the Neutron service. It used legacy security group APIs to manage access to instances, and there was no proper support for custom networking.
+
+Starting from Apache jclouds 2.1.0, an integration with OpenStack Neutron will be provided and users will be able to configure their Nova APIs to interact with a custom Neutron deployment.
+<!-- more -->
+
+To achieve this, users will be able to use the recent **context linking** feature, where APIs and providers that have dependencies between them can be *linked* so they can call each other where needed.
+
+## Linking OpenStack Nova to Neutron
+
+Links between APIs and providers are done at *context* level. This isolates each individual context and allows users to configure an independent set of properties for each one without overlapping issues. The `ContextLinking.linkContext` and `ContextLinking.linkView` helper methods can be used to easily link one context or view to another.
+
+The following example shows how to link an OpenStack Nova API to a Neutron API context, to leverage Neutron features when provisioning instances with Nova:
+
+{% highlight java %}
+// Create the connection to OpenStack Neutron
+ApiContext<NeutronApi> neutronCtx = ContextBuilder.newBuilder("openstack-neutron")
+   .endpoint("http://localhost/identity/v3/")
+   .credentials("domain:user", "password")
+   .overrides(neutronProperties)
+   .modules(ImmutableSet.of(new SLF4JLoggingModule()))
+   .build();
+
+// Create the connection to OpenStack nova and link it to Neutron
+NovaApi nova = ContextBuilder.newBuilder("openstack-nova")
+   .endpoint("http://localhost/identity/v3/")
+   .credentials("domain:user", "password")
+   .overrides(novaProperties)
+   .modules(ImmutableSet.of(
+               ContextLinking.linkContext(neutronCtx),
+               new SLF4JLoggingModule()))
+   .buildApi(NovaApi.class);
+{% endhighlight %}
+
+With this configuration the `nova` API is configured to use the linked `neutron` for all networking operations.
```

Comment:
Because of the natural meaning of the APIs. Neutron is just a networking API, whilst Nova is the compute API that also performs networking operations when creating servers. It is nova the one that logically uses neutron, not the other way. I guess this would be what most users will already assume?

Reply:
> I guess this would be what most users will already assume?

I was just asking because "link" is bidirectional as a verb (as opposed to, say, "use") - it could mean "uses", as in this case, or "connected to", which could be the other way around (i.e. "Neutron connected to Nova").

Would it make sense, in the description of context linking, to extend the current sentence "...helper methods can be used to easily link one context or view to another." to something like "Links are specified in the 'using' context and point to the context(s) that it uses."?

Reply:
@andreaturli Curious to hear your thoughts on this one, too ;-)

Reply:
Sorry @nacx and @demobox for late comment, I think the current wording is good enough.
I'd prefer to see comments on the blog posts asking for more clarity rather than try to come up with the perfect sentence :)

I think there is enough material to add another blog post on Openstack specifically on how jclouds can manage security groups and floating ips from nova and neutron depending on the linked context.

Reply:
> I'd prefer to see comments on the blog posts asking for more clarity rather than try to come up with the perfect sentence :)

A fair point, although of course we are unlikely to know how many people are confused but do *not* leave comments ;-)